clippy_lints/src/recursive_format_impl.rs

```diff
@@ -0,0 +1,269 @@
+}
+
+impl RecursiveFormatImpl {
+    pub fn new() -> Self {
+        Self {
+            format_trait_impl: None,
+            self_hir_id: None,
+        }
+    }
+}
+
+impl_lint_pass!(RecursiveFormatImpl => [RECURSIVE_FORMAT_IMPL]);
+
+impl LateLintPass<'_> for RecursiveFormatImpl {
+    fn check_item(&mut self, cx: &LateContext<'_>, item: &Item<'_>) {
+        if let Some(format_trait_impl) = is_format_trait_impl(cx, item) {
+            self.format_trait_impl = Some(format_trait_impl);
+        }
+    }
+
+    fn check_item_post(&mut self, cx: &LateContext<'_>, item: &Item<'_>) {
+        // Assume no nested Impl of Debug and Display within eachother
+        if is_format_trait_impl(cx, item).is_some() {
+            self.format_trait_impl = None;
+            self.self_hir_id = None;
+        }
+    }
+
+    fn check_impl_item(&mut self, cx: &LateContext<'_>, impl_item: &ImplItem<'_>) {
+        if_chain! {
+            // If we are in Display or Debug impl, then get hir_id for self in method impl - i.e. fmt(&self)
+            if self.format_trait_impl.is_some();
+            if let ImplItemKind::Fn(.., body_id) = &impl_item.kind;
+            let body = cx.tcx.hir().body(*body_id);
+            if !body.params.is_empty();
+            then {
+                let self_param = &body.params[0];
+                self.self_hir_id = Some(self_param.pat.hir_id);
+            }
+        }
+    }
+
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        if let Some(self_hir_id) = self.self_hir_id {
+            match self.format_trait_impl {
+                Some(ImplTrait::Display) => {
+                    check_to_string_in_display(cx, expr, self_hir_id);
+                    check_self_in_format_args(cx, expr, self_hir_id, ImplTrait::Display);
+                },
+                Some(ImplTrait::Debug) => {
+                    check_self_in_format_args(cx, expr, self_hir_id, ImplTrait::Debug);
+                },
+                None => {},
+            }
+        }
+    }
+}
+
+fn check_to_string_in_display(cx: &LateContext<'_>, expr: &Expr<'_>, self_hir_id: HirId) {
+    if_chain! {
+        // Get the hir_id of the object we are calling the method on
+        if let ExprKind::MethodCall(path, _, [ref self_arg, ..], _) = expr.kind;
+        // Is the method to_string() ?
+        if path.ident.name == sym!(to_string);
+        // Is the method a part of the ToString trait? (i.e. not to_string() implemented
+        // separately)
+        if let Some(expr_def_id) = cx.typeck_results().type_dependent_def_id(expr.hir_id);
+        if is_diag_trait_item(cx, expr_def_id, sym::ToString);
+        // Is the method is called on self
+        if path_to_local_id(self_arg, self_hir_id);
+        then {
+            span_lint(
+                cx,
+                RECURSIVE_FORMAT_IMPL,
+                expr.span,
+                "using `to_string` in `fmt::Display` implementation might lead to infinite recursion",
```

Comment:
I think we can be confident about it leading to infinite recursion. Unless you have a counterexample in mind?
```suggestion
                "using `self.to_string` in `fmt::Display` implementation will cause infinite recursion",
```